### PR TITLE
Added customUrlHandler feature as described in #60

### PR DESCRIPTION
**Pull request by Mkots**
And enabled use of all available CPU cores for Linux compilation

### CODE REVIEW

File: src/session.ts [hunk 1]
```diff
@@ -665,6 +689,29 @@ export class Session extends EventEmitter {
 
     let url = this.url
 
+    if (this.customUrlHandler && !this.url) {
+      const customUrlHandlerPath = path.resolve(
+        process.cwd(),
+        this.customUrlHandler,
+      )
+      if (!fs.existsSync(customUrlHandlerPath)) {
+        console.error(`Custom error handler not found: ${customUrlHandlerPath}`)
```
Comment (vpalmisano) [inline src/session.ts]:
```suggestion
        log.error(`Custom error handler not found: "${customUrlHandlerPath}"`)
```

File: src/session.ts [hunk 2]
```diff
@@ -665,6 +689,29 @@ export class Session extends EventEmitter {
...
+        process.exit(1)
```
Comment (vpalmisano) [inline src/session.ts]:
```suggestion
        throw new Error(`Custom error handler not found: "${customUrlHandlerPath}"`)
```

File: src/session.ts [hunk 3]
```diff
@@ -665,6 +689,29 @@ export class Session extends EventEmitter {
...
+      }
+      const customUrlHandler = (
+        await import(/* webpackIgnore: true */ customUrlHandlerPath)
+      ).default
+
+      url = customUrlHandler({
```
Comment (vpalmisano) [inline src/session.ts]: I would mark the `customUrlHandler` function as async, just to cover all the possible use cases.

File: src/session.ts [hunk 4]
```diff
@@ -257,6 +259,22 @@ export class Session extends EventEmitter {
   readonly url: string
   /** The url query. */
   readonly urlQuery: string
+  /**
+   * The custom URL handler. This is the path to the JavaScript file containing the function.
```
Comment (vpalmisano) [inline src/session.ts]:
```suggestion
   * The custom URL handler. This is the path to a JavaScript module (.mjs) exporting the function.
```

File: src/session.ts [hunk 5]
```diff
@@ -257,6 +259,22 @@ export class Session extends EventEmitter {
...
+   * @typedef {Object} CustomUrlHandler
+   * @property {string} id - The identifier for the URL.
+   * @property {string} sessions - The number of sessions.
+   * @property {string} tabIndex - The index of the current tab.
+   * @property {string} tabsPerSession - The number of tabs per session.
+   * @property {string} index - The index for the URL.
+   * @property {string} pid - The process identifier for the URL.
+   *
+   * @type {string} path - The path to the JavaScript file containing the function:
+   *   (params: CustomUrlHandler) => string
```
Comment (vpalmisano) [inline src/session.ts]: Using async:
```suggestion
   *   (params: CustomUrlHandler) => Promise<string>
```

**Review by vpalmisano — CHANGES_REQUESTED** [review]

File: package.json [hunk 6]
```diff
@@ -36,7 +36,7 @@
     "lint": "eslint --resolve-plugins-relative-to path=. src scripts examples",
     "lint:fix": "yarn lint --fix",
     "docs": "ts-node src/generate-config-docs.ts && typedoc",
-    "build:linux": "nexe --verbose --build -i app.min.js -o webrtcperf -r 'scripts/*' -r package.json && gzip -9 -c webrtcperf > webrtcperf_${npm_package_version}_linux.gz",
+    "build:linux": "nexe --make=\"-j$(nproc 2> /dev/null || echo 1)\" --verbose --build -i app.min.js -o webrtcperf -r 'scripts/*' -r package.json && gzip -9 -c webrtcperf > webrtcperf_${npm_package_version}_linux.gz",
```
Comment (vpalmisano) [inline package.json:39]: Ok, but next time do it in a separate PR ;)

File: src/config.ts [hunk 7]
```diff
@@ -41,6 +41,17 @@ session, \`$i\` the tab absolute index.`,
     env: 'URL_QUERY',
     arg: 'url-query',
   },
+  customUrlHandler: {
+    doc: `The path to custom page url handler; the following \
```
Comment (vpalmisano) [inline src/config.ts]: Maybe we need to change the doc comment here, because those variables are actually passed to the exported function.

**Review by vpalmisano — COMMENTED** [review]
I'm thinking about reusing the `--url` config for both cases:
- url.startsWith('http') -> current usage
- else, try to load the file at `url` path.
What do you think?

File: src/session.ts [hunk 8]
```diff
@@ -665,6 +689,29 @@ export class Session extends EventEmitter {
...
+    if (this.customUrlHandler && !this.url) {
+      const customUrlHandlerPath = path.resolve(
+        process.cwd(),
+        this.customUrlHandler,
+      )
+      if (!fs.existsSync(customUrlHandlerPath)) {
+        console.error(`Custom error handler not found: ${customUrlHandlerPath}`)
+        process.exit(1)
+      }
+      const customUrlHandler = (
```
Comment (vpalmisano) [inline src/session.ts]: I would move this in the constructor, assigning the imported function to the class instance.

File: src/session.ts [hunk 9]
```diff
@@ -665,6 +711,32 @@ export class Session extends EventEmitter {
...
+        process.cwd(),
+        this.customUrlHandler,
+      )
+      if (!fs.existsSync(customUrlHandlerPath)) {
+        log.error(`Custom error handler not found: "${customUrlHandlerPath}"`)
+        throw new Error(
+          `Custom error handler not found: "${customUrlHandlerPath}"`,
+        )
+      }
+      this.customUrlHandlerFn = (
```
Comment (vpalmisano) [inline src/session.ts:725]: Can we move this assignment into the class constructor?

Comment (Mkots) [reply src/session.ts:725]: We can do it, but it would be slightly complicated because importing is an asynchronous process, so it will increase code complications in the session class. 
If you find this change important, I can figure something out, but I don't understand the necessity of this.

**Review by vpalmisano — APPROVED** [review]
